### In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

- The dataset can be downloaded from here :- https://data.world/data-society/used-cars-data

- the original dataset is already cleaned the same dataset used in this project is taken from the dataquest which consists of 50000 rows and 20 columns which is messy and requires some attention.

- the dataset will be available in my github as autos.csv

In [92]:
## The aim of this project is to clean the data and analyze the included used car listings.


The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [93]:
# importing the libraries
import numpy as np
import pandas as pd

In [94]:
# Reading the dataset
autos = pd.read_csv('autos.csv',encoding="Latin-1")

In [95]:
# checking the dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [96]:
# Getting the information of the dataset
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [97]:
# Getting the initial rows of the dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfle...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Observations
- dataset consists of 20 columns of which most are strings
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

## 2.Cleaning column names
we will convert convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive

In [98]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

In [99]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price  ab_test vehicle_type  registration_year   gear_box  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  not_repaired_damage           ad_created  num_pictures  postal_code  \
0                nein  2016-03-26 00:00:00             0        79588   
1                nein  2016-04-04 00:00:00             0        71034   
2                nein  2016-03-26 00:00:00             0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

## 3.Initial Exploration and Cleaning


In [100]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  num_pictures   postal_code  \
count                40171                50000       50000.0  50000.000000   
unique                   2                   76           NaN           NaN   
top                   nein  2016-04-03 00:00:00           NaN           NaN   
freq                 35232                 1946           NaN           NaN   
mean                   NaN                  NaN           0.0  50813.627300   
std                    NaN                  NaN           0.0  25779.747957   
min                    NaN                  NaN           0.0   1067.000000   
25%                    NaN                  NaN           0.0  30451.000000   
50%                    NaN                  NaN           0.0  49577.000000   
75%                    NaN                  NaN           0.0  71540.000000   
max                    NaN                  NaN           0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

- from the dataset it can be seen that price and odometer columns are numeric but both are stored as an object



- odometer is a instrument used for measuring the distance travelled by the vehicle and the column will be renamed from odometer to odometer_km

In [101]:
# convert price column to numeric datatype and replace the non numeric characters from the dataset.
# autos.loc[:,"price"] = 
autos.loc[:,"price"].value_counts()

$0          1421
$500         781
$1,500       734
$2,500       643
$1,000       639
            ... 
$7,449         1
$14,899        1
$29,890        1
$151,990       1
$59,850        1
Name: price, Length: 2357, dtype: int64

In [102]:
autos.loc[:,"price"] = autos.loc[:,"price"].str.replace("$","").str.replace(",","").astype(int)


autos.loc[:,"price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

In [103]:
# convert odometer column to numeric datatype and replace the non numeric characters from the dataset.
autos.loc[:,"odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [104]:
autos.loc[:,"odometer"] = autos.loc[:,"odometer"].str.replace("km","").str.replace(",","").astype(int)

autos.loc[:,"odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int32

In [105]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   date_crawled         50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offer_type           50000 non-null  object
 4   price                50000 non-null  int32 
 5   ab_test              50000 non-null  object
 6   vehicle_type         44905 non-null  object
 7   registration_year    50000 non-null  int64 
 8   gear_box             47320 non-null  object
 9   power_ps             50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  int32 
 12  registration_month   50000 non-null  int64 
 13  fuel_type            45518 non-null  object
 14  brand                50000 non-null  object
 15  not_repaired_damage  40171 non-null  object
 16  ad_c

In [106]:
# renaming the column
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

In [107]:
autos.loc[:,"seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [108]:
autos.loc[:,"offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [109]:
autos.loc[:,"num_pictures"].value_counts()

0    50000
Name: num_pictures, dtype: int64

## Observations
- the seller, offertype has the same values for every rows and therefore it does not feel significant
- the num pictures column also seems a little odd as it has 0 value for all the 500000 rows

In [110]:
autos= autos.drop(["num_pictures","seller","offer_type"],axis=1)

In [111]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   date_crawled         50000 non-null  object
 1   name                 50000 non-null  object
 2   price                50000 non-null  int32 
 3   ab_test              50000 non-null  object
 4   vehicle_type         44905 non-null  object
 5   registration_year    50000 non-null  int64 
 6   gear_box             47320 non-null  object
 7   power_ps             50000 non-null  int64 
 8   model                47242 non-null  object
 9   odometer_km          50000 non-null  int32 
 10  registration_month   50000 non-null  int64 
 11  fuel_type            45518 non-null  object
 12  brand                50000 non-null  object
 13  not_repaired_damage  40171 non-null  object
 14  ad_created           50000 non-null  object
 15  postal_code          50000 non-null  int64 
 16  last

## 4. Exploring the Odometer and Price Columns

In [112]:
autos.loc[:,"odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [113]:
# there are more high mileage vehicles then low mileage vehicles and all the values are seem to be rounded

In [114]:
print(autos.loc[:,"price"].unique().shape)

print("\r")

print(autos.loc[:,"price"].value_counts())

print("\r")

print(autos.loc[:,"price"].describe())

(2357,)

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [115]:
# again the values in the price column are rounded
# there are 1421 cars which has 0 value and this accounts for approximately 2% of the data and can be deleted.
# the maximum price seems to be 100 million dollars which seems to be unrealistic so lets look the data again

In [116]:
autos.loc[:,"price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [117]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

### Observation

- Since Ebay is an auction site the bid can be as low as 1$ and we will keep the prices till 350000 as after that prices seem to be increasing drastically to some unrealistic numbers


- There are a number of listings with prices below 30 including about 1,500 at 0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

In [118]:
# selecting rows between 0 and 351000
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## 5.Exploring the date columns

- There are 5 columns that should represent date values.
- Some of these columns were created by the crawler, some came from the website itself.


In [119]:
# Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas.

In [120]:
# Let's first understand how the values in the three string columns are formatted. 
# These columns all represent full timestamp values, like so:
autos.loc[:,["date_crawled","last_seen","ad_created"]].head(5)

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

In [121]:
# starting 10 characters of the date crawled column is the date

autos.loc[:,"date_crawled"]= autos.loc[:,"date_crawled"].str[:10]

In [122]:
autos.loc[:,"date_crawled"].value_counts(normalize=True, dropna=False).sort_index(),3

(2016-03-05    0.025327
 2016-03-06    0.014043
 2016-03-07    0.036014
 2016-03-08    0.033296
 2016-03-09    0.033090
 2016-03-10    0.032184
 2016-03-11    0.032575
 2016-03-12    0.036920
 2016-03-13    0.015670
 2016-03-14    0.036549
 2016-03-15    0.034284
 2016-03-16    0.029610
 2016-03-17    0.031628
 2016-03-18    0.012911
 2016-03-19    0.034778
 2016-03-20    0.037887
 2016-03-21    0.037373
 2016-03-22    0.032987
 2016-03-23    0.032225
 2016-03-24    0.029342
 2016-03-25    0.031607
 2016-03-26    0.032204
 2016-03-27    0.031092
 2016-03-28    0.034860
 2016-03-29    0.034099
 2016-03-30    0.033687
 2016-03-31    0.031834
 2016-04-01    0.033687
 2016-04-02    0.035478
 2016-04-03    0.038608
 2016-04-04    0.036487
 2016-04-05    0.013096
 2016-04-06    0.003171
 2016-04-07    0.001400
 Name: date_crawled, dtype: float64,
 3)

In [123]:
autos.loc[:,"date_crawled"].value_counts(normalize=True, dropna=False).sort_values()

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

In [124]:
autos.loc[:,"date_crawled"].isnull().sum() # there are no null values in the date crawled column

0

In [125]:
autos.loc[:,"ad_created"]= autos.loc[:,"ad_created"].str[:10]

In [126]:
autos.loc[:,"ad_created"].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [127]:
autos.loc[:,"ad_created"].isnull().sum()

0

In [128]:
autos.loc[:,"last_seen"]= autos.loc[:,"last_seen"].str[:10]


In [129]:
autos.loc[:,"last_seen"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [130]:
autos.loc[:,"ad_created"].isnull().sum()

0

In [131]:
# checking the values for all the columns from earliest to latest

autos.loc[:,"ad_created"].unique()

array(['2016-03-26', '2016-04-04', '2016-03-12', '2016-04-01',
       '2016-03-21', '2016-03-20', '2016-03-16', '2016-03-22',
       '2016-03-14', '2016-03-31', '2016-03-23', '2016-03-29',
       '2016-03-17', '2016-03-05', '2016-03-06', '2016-03-28',
       '2016-03-10', '2016-04-03', '2016-03-19', '2016-04-02',
       '2016-03-15', '2016-04-05', '2016-03-11', '2016-03-07',
       '2016-03-08', '2016-03-27', '2016-03-09', '2016-03-25',
       '2016-03-18', '2016-03-30', '2016-03-24', '2016-03-13',
       '2016-04-06', '2016-03-04', '2016-04-07', '2016-02-24',
       '2016-03-02', '2016-02-29', '2016-01-03', '2015-11-10',
       '2016-03-01', '2016-03-03', '2016-02-28', '2016-01-27',
       '2016-02-19', '2016-02-25', '2016-02-23', '2016-02-14',
       '2016-02-12', '2016-01-29', '2016-02-05', '2016-01-22',
       '2016-02-01', '2016-02-02', '2016-02-17', '2016-02-21',
       '2015-08-10', '2015-06-11', '2016-02-11', '2016-01-14',
       '2016-01-10', '2016-02-08', '2015-12-05', '2016-

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [132]:
# checking the description of registration_year column
autos.loc[:,"registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

there seems to be some error in the registration year as the minimum year is around 1000 long before the car was invented and maximum is around 9999 many years into the future.

## 6. Dealing with Incorrect Registration Year Data

- Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate.

In [133]:
((~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0])*100

3.8793369710697

roughly there are only 4% of vehicles not in this range. therefore the values outside this can be removed.

In [134]:
autos = autos[autos.loc[:,"registration_year"].between(1900,2016)]

In [135]:
autos.loc[:,"registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [136]:
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

it appears that most of the vehicles were registered in the past 20 years.

## 7. Exploring Price by Brand

In [137]:
autos.loc[:,"brand"].value_counts(normalize=True).head(20)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
Name: brand, dtype: float64

In [138]:
common_brand = autos.loc[:,"brand"].unique()

In [139]:
common_brand

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [140]:
brand_mean_prices={}

for row in common_brand:
    brand_only = autos[autos.loc[:,"brand"]== row]
    mean_price= brand_only.loc[:,"price"].mean()
    brand_mean_prices[row] = int(mean_price)

In [141]:
brand_mean_prices

{'peugeot': 3094,
 'bmw': 8332,
 'volkswagen': 5402,
 'smart': 3580,
 'ford': 3749,
 'chrysler': 3465,
 'seat': 4397,
 'renault': 2474,
 'mercedes_benz': 8628,
 'audi': 9336,
 'sonstige_autos': 12338,
 'opel': 2975,
 'mazda': 4112,
 'porsche': 45643,
 'mini': 10613,
 'toyota': 5167,
 'dacia': 5915,
 'nissan': 4743,
 'jeep': 11650,
 'saab': 3211,
 'volvo': 4946,
 'mitsubishi': 3394,
 'jaguar': 11635,
 'fiat': 2813,
 'skoda': 6368,
 'subaru': 3953,
 'kia': 5982,
 'citroen': 3779,
 'chevrolet': 6684,
 'hyundai': 5365,
 'honda': 4107,
 'daewoo': 1049,
 'suzuki': 4096,
 'trabant': 1790,
 'land_rover': 19108,
 'alfa_romeo': 4087,
 'lada': 2688,
 'rover': 1602,
 'daihatsu': 1636,
 'lancia': 3376}

In [142]:
autos.head(3)

date_crawled                                        name  price  ab_test  \
0   2016-03-26            Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   
1   2016-04-04  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500  control   
2   2016-03-26                  Volkswagen_Golf_1.6_United   8990     test   

  vehicle_type  registration_year   gear_box  power_ps   model  odometer_km  \
0          bus               2004    manuell       158  andere       150000   
1    limousine               1997  automatik       286     7er       150000   
2    limousine               2009    manuell       102    golf        70000   

   registration_month fuel_type       brand not_repaired_damage  ad_created  \
0                   3       lpg     peugeot                nein  2016-03-26   
1                   6    benzin         bmw                nein  2016-04-04   
2                   7    benzin  volkswagen                nein  2016-03-26   

   postal_code   last_seen  
0        79588  2016-04-06  
1        71034  2016-04-06  
2        35394  2016-04-06

In [143]:
brand_mean_mileage={}

for row in common_brand:
    brand_only = autos[autos.loc[:,"brand"]== row]
    mean_mileage= brand_only.loc[:,"odometer_km"].mean()
    brand_mean_mileage[row] = int(mean_mileage)

In [144]:
brand_mean_mileage

{'peugeot': 127153,
 'bmw': 132572,
 'volkswagen': 128707,
 'smart': 99326,
 'ford': 124266,
 'chrysler': 132378,
 'seat': 121131,
 'renault': 128071,
 'mercedes_benz': 130788,
 'audi': 129157,
 'sonstige_autos': 89956,
 'opel': 129310,
 'mazda': 124464,
 'porsche': 96853,
 'mini': 88105,
 'toyota': 115944,
 'dacia': 84268,
 'nissan': 118330,
 'jeep': 127122,
 'saab': 144415,
 'volvo': 138067,
 'mitsubishi': 126575,
 'jaguar': 124178,
 'fiat': 117121,
 'skoda': 110848,
 'subaru': 125300,
 'kia': 112530,
 'citroen': 119694,
 'chevrolet': 99191,
 'hyundai': 106442,
 'honda': 122172,
 'daewoo': 121642,
 'suzuki': 108465,
 'trabant': 54538,
 'land_rover': 118010,
 'alfa_romeo': 131338,
 'lada': 83518,
 'rover': 137661,
 'daihatsu': 116410,
 'lancia': 121900}

In [145]:
# Converting both the series into a series object
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

peugeot            3094
bmw                8332
volkswagen         5402
smart              3580
ford               3749
chrysler           3465
seat               4397
renault            2474
mercedes_benz      8628
audi               9336
sonstige_autos    12338
opel               2975
mazda              4112
porsche           45643
mini              10613
toyota             5167
dacia              5915
nissan             4743
jeep              11650
saab               3211
volvo              4946
mitsubishi         3394
jaguar            11635
fiat               2813
skoda              6368
subaru             3953
kia                5982
citroen            3779
chevrolet          6684
hyundai            5365
honda              4107
daewoo             1049
suzuki             4096
trabant            1790
land_rover        19108
alfa_romeo         4087
lada               2688
rover              1602
daihatsu           1636
lancia             3376
dtype: int64


In [146]:
bmm_series = pd.Series(brand_mean_mileage)
print(bmm_series)

peugeot           127153
bmw               132572
volkswagen        128707
smart              99326
ford              124266
chrysler          132378
seat              121131
renault           128071
mercedes_benz     130788
audi              129157
sonstige_autos     89956
opel              129310
mazda             124464
porsche            96853
mini               88105
toyota            115944
dacia              84268
nissan            118330
jeep              127122
saab              144415
volvo             138067
mitsubishi        126575
jaguar            124178
fiat              117121
skoda             110848
subaru            125300
kia               112530
citroen           119694
chevrolet          99191
hyundai           106442
honda             122172
daewoo            121642
suzuki            108465
trabant            54538
land_rover        118010
alfa_romeo        131338
lada               83518
rover             137661
daihatsu          116410
lancia            121900


In [147]:
# Creating a Dataframe from the first series
df_price = pd.DataFrame(bmp_series,columns=["mean_price"])

In [148]:
df_price["mean_mileage"] = bmm_series

In [149]:
df_price.sort_values('mean_price',ascending=False)

mean_price  mean_mileage
porsche              45643         96853
land_rover           19108        118010
sonstige_autos       12338         89956
jeep                 11650        127122
jaguar               11635        124178
mini                 10613         88105
audi                  9336        129157
mercedes_benz         8628        130788
bmw                   8332        132572
chevrolet             6684         99191
skoda                 6368        110848
kia                   5982        112530
dacia                 5915         84268
volkswagen            5402        128707
hyundai               5365        106442
toyota                5167        115944
volvo                 4946        138067
nissan                4743        118330
seat                  4397        121131
mazda                 4112        124464
honda                 4107        122172
suzuki                4096        108465
alfa_romeo            4087        131338
subaru                3953        125300
citroen               3779        119694
ford                  3749        124266
smart                 3580         99326
chrysler              3465        132378
mitsubishi            3394        126575
lancia                3376        121900
saab                  3211        144415
peugeot               3094        127153
opel                  2975        129310
fiat                  2813        117121
lada                  2688         83518
renault               2474        128071
trabant               1790         54538
daihatsu              1636        116410
rover                 1602        137661
daewoo                1049        121642

## Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

## Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
- How much cheaper are cars with damage than their non-damaged counterparts?

In [150]:
# Identifying categorical data that uses german words, translate them and map the values to their english counterparts
autos.loc[:,"gear_box"].value_counts()


manuell      34715
automatik     9856
Name: gear_box, dtype: int64

In [151]:
gearbox = {"manuell":"manual","automatik":"automatic"}

In [152]:
autos.loc[:,"gear_box"] = autos.loc[:,"gear_box"].map(gearbox)

In [153]:
autos.loc[:,"gear_box"]

0           manual
1        automatic
2           manual
3        automatic
4           manual
           ...    
49995    automatic
49996       manual
49997    automatic
49998       manual
49999       manual
Name: gear_box, Length: 46681, dtype: object

In [154]:
autos.loc[:,"vehicle_type"].value_counts()

limousine     12598
kleinwagen    10585
kombi          8930
bus            4031
cabrio         3016
coupe          2462
suv            1965
andere          390
Name: vehicle_type, dtype: int64

 - kleinwagen is a german word for small car
 - cabrio is convertible
 - kombi is minibus